### PR TITLE
feat(builder): add ssh key variable to support private repositories

### PR DESCRIPTION
**Pull request by rvaralda**
Just reopened because a merge mess.

Ref: #3259

### CODE REVIEW

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by rvaralda** [comment]
@bacongobbler There is the whole package, check if it is ok.

**Comment by rvaralda** [comment]
Just to update.
I have tested the following cases:

`$ deis config:set SSH_KEY=~/.ssh/<priv-key>` :+1: 

`$ deis config:set SSH_KEY="-----BEGIN RSA PRIVATE KEY-----(...)"` :+1: 

`$ deis config:set SSH_KEY=~/wrong-dir/<somefile>`
Got the error: `Could not parse SSH private key ~/wrong-dir/<somefile>`

`deis config:set SSH_KEY="wrong content"`
Got the error: `Could not parse SSH private key wrong content`

**Comment by bacongobbler** [comment]
Sounds good!

Tenatively setting this for v1.5 release testing.

**Comment by carmstrong** [comment]
Code LGTM at first pass. Had to think this through a bit, but now I get it.

**Comment by bacongobbler** [comment]
Jenkins, add to whitelist.

**Comment by carmstrong** [comment]
@rvaralda Looks like it's failing flake tests with:
```
./deis.py:1054:80: F821 undefined name 'ssh_key'
```

File: docs/using_deis/using-buildpacks.rst [hunk 1]
```diff
@@ -27,6 +27,29 @@ Use ``deis create`` to create an application on the :ref:`controller`.
...
+
+
+Using private key file:
+
+.. code-block:: console
+
+    $ deis config:set SSH_KEY=/home/user/.ssh/id_rsa
+
+
+You also can use raw key instead private key file:
```
Comment (mboersma) [inline docs/using_deis/using-buildpacks.rst]: Could you change this line to
> You can also use a raw key instead of a private key file:

**Comment by mboersma** [comment]
@rvaralda could you `git rebase` and squash these two commits into one with a commit message like:
```feat(builder): add ssh key variable to support private repositories```

**Comment by mboersma** [comment]
Excellent, thanks! Code LGTM.

**Comment by mboersma** [comment]
I've been trying to recreate a situation to test this, but I'm failing. I cloned heroku-buildpack-go and put it in a private GitHub repo, but after setting `BUILDPACK_URL` and `SSH_KEY` I get this:
```console
$ git push deis master
Counting objects: 28, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (23/23), done.
Writing objects: 100% (28/28), 10.04 KiB | 0 bytes/s, done.
Total 28 (delta 7), reused 0 (delta 0)
-----> Fetching custom buildpack
fatal: could not read Username for 'https://github.com': No such device or address
To ssh://git@deis.local3.deisapp.com:2222/queens-rucksack.git
 * [new branch]      master -> master
```
@rvaralda any advice on how I can verify this PR?

**Comment by rvaralda** [comment]
Hey @mboersma,

It will try to reach vi ssh and match the key file to auth.
Try add buildpack url like this: git@github.com:deis/private-repo.git

**Comment by mboersma** [comment]
> add buildpack url like this: git@github.com:deis/private-repo.git

Of course--that works. Also it helps if the `SSH_KEY` is actually uploaded to GitHub. :smile: 

I pushed heroku-buildpack-go to a private repository and set it as `BUILDPACK_URL`, then confirmed that `git push deis master` only succeeds after `deis config:set SSH_KEY=`. I tested the same variations mentioned above. LGTM!

**Comment by rvaralda** [comment]
great :+1:

**Comment by bacongobbler** [comment]
lgtm! Review seems sound, though maybe we could elaborate more in the docs on how one can use a private repository. Some people may not know how to set this up. Maybe a blog post? :)

**Comment by mboersma** [comment]
> Maybe a blog post?

Good idea, it could be short and sweet.

**Pull request merged**